quic/s2n-quic-platform/src/message.rs

```diff
@@ -25,7 +25,51 @@ pub mod default {
     }
 }
 
-pub type Storage = Pin<Box<[UnsafeCell<u8>]>>;
+/// Tracks allocations of message ring buffer state
+pub struct Storage {
+    ptr: NonNull<UnsafeCell<MaybeUninit<u8>>>,
+    layout: Layout,
+}
+
+impl Storage {
+    #[inline]
+    pub fn new(layout: Layout) -> Self {
+        unsafe {
+            let ptr = alloc::alloc::alloc_zeroed(layout);
+            let ptr = ptr as *mut UnsafeCell<MaybeUninit<u8>>;
+            let ptr = NonNull::new(ptr).expect("could not allocate message storage");
+            Self { layout, ptr }
+        }
+    }
+
+    #[inline]
+    pub(crate) fn as_ptr(&self) -> *mut UnsafeCell<MaybeUninit<u8>> {
+        self.ptr.as_ptr().cast()
+    }
+
+    /// Asserts that the pointer is in bounds of the allocation
+    #[inline]
+    pub(crate) fn check_bounds<T: Sized>(&self, ptr: *mut T) {
+        let expected_end_ptr = unsafe {
+            // Safety: pointer is allocated with the self.layout
+            self.as_ptr().add(self.layout.size())
+        };
+        let actual_end_ptr = unsafe {
+            // increment the pointer by one to ensure it can be dereferenced
+            ptr.add(1) as *mut UnsafeCell<MaybeUninit<u8>>
+        };
+        debug_assert!(expected_end_ptr >= actual_end_ptr);
```

Comment:
I assume this is mostly a miri check? I think at runtime (technically with miri but it's much less likely) it would be sound to lower this to a check that layout
size > 1, because add is UB if it wraps or goes outside that one allocation. And size > 1 isn't a useful check. wrapping_add or casting ptr to a usize is probably what I'd do.

Reply:
Miri already checks that pointers are within bounds, right? This is just a runtime sanity check for all of the pointer arithmetic to make sure it's all within the allocation. But it's probably good to avoid `.add` since that wraps.

Reply:
Oh, I misread this code - I thought `ptr` and `self.as_ptr()` where the same but they're distinct here. Yeah, this should be OK but I think using wrapping_add is safer. As-is this function is unsound in the general case since ptr.add(1) may not be in-bounds to an allocation.

Reply:
I ended up doing a range with no adding in this function itself. Using a range should be able to detect wrapping as well, since it checks the start pointer.